src/rviz/default_plugin/markers/mesh_resource_marker.cpp

```diff
@@ -192,49 +179,57 @@ void MeshResourceMarker::onNewMessage(const MarkerConstPtr& old_message, const M
       entity_->setMaterial(default_material);
     }
 
-    update_color = true;
+    update_color = !(new_message->mesh_use_embedded_materials && r == 0 && g == 0 && b == 0 && a == 0);
 
     handler_.reset(new MarkerSelectionHandler(this, MarkerID(new_message->ns, new_message->id), context_));
     handler_->addTrackedObject(entity_);
   }
   else
   {
     // underlying mesh resource has not changed but if the color has
-    //  then we need to update the materials color
-    if (new_message->mesh_use_embedded_materials == false
-       && (!old_message
-        || old_message->mesh_use_embedded_materials == true
-        || old_message->color.r != new_message->color.r
-        || old_message->color.g != new_message->color.g
-        || old_message->color.b != new_message->color.b
-        || old_message->color.a != new_message->color.a))
+    // then we need to update the materials color
+    if (!old_message
+        || old_message->color.r != r
+        || old_message->color.g != g
+        || old_message->color.b != b
+        || old_message->color.a != a)
     {
       update_color = true;
     }
   }
 
   // update material color
-  //  if the mesh_use_embedded_materials is true and color is non-zero
-  //  then the color will be used to tint the embedded materials
   if (update_color)
   { 
-    if( new_message->mesh_use_embedded_materials && r == 0 && g == 0 && b == 0 && a == 0 )
-    {
-    blending = Ogre::SBT_REPLACE;
-    depth_write = true;
-    r = 1; g = 1; b = 1; a = 1;
-    }
-    S_MaterialPtr::iterator material_it;
-    for (material_it = materials_.begin(); material_it != materials_.end(); material_it++)
+    bool depth_write = a >= 0.9998;
+    Ogre::SceneBlendType blending = depth_write ? Ogre::SBT_REPLACE : Ogre::SBT_TRANSPARENT_ALPHA;
+    bool tinting = new_message->mesh_use_embedded_materials;
+
+    for (const Ogre::MaterialPtr& material : materials_)
     {
-      Ogre::Technique* technique = (*material_it)->getTechnique(0);
-      technique->setAmbient( r*0.5, g*0.5, b*0.5 );
-      technique->setDiffuse( r, g, b, a );
-      technique->setSceneBlending( blending );
-      technique->setDepthWriteEnabled( depth_write );
-      technique->setLightingEnabled( true );
+      Ogre::Technique* technique = material->getTechnique(0);
+      Ogre::Pass* pass0 = technique->getPass(0);
+      Ogre::Pass* passT = technique->getPass(technique->getNumPasses()-1);
+      if (tinting)
+      {
+        // modify material's original color to use given alpha value
+        Ogre::ColourValue color = pass0->getDiffuse();
+        color.a = a;
```

Comment:
we faced a problem in gazebo before in which the mesh's material is also semi-transparent. We ended up just combining the alpha values, e.g. `color.a *= a`, so the alpha effect is multiplied. Overriding the value is also fine as long as the user knows what to expect

Reply:
@iche033, thanks for reviewing. 
Multiplying the values is dangerous as multiple updates of the marker's alpha value will continuously decrease the effective alpha value. (We don't have a memory of the original alpha value of the mesh's material.)
To allow to increase the alpha value again, we need to overwrite as proposed.
Hence, the first time the user provides a custom alpha value via the marker's colors, this will take precedence over the mesh's alpha values.